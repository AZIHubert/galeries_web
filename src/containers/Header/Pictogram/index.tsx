import * as React from 'react';
import { CSSTransition } from 'react-transition-group';
import styled from 'styled-components';

interface ContainerI {
  marginLeft?: number;
  marginRight?: number;
}

interface PictogramI {
  hoverPictogram: React.ComponentType;
  marginLeft?: number;
  marginRight?: number;
  pictogram: React.ComponentType;
}

const Container = styled.div<ContainerI>`
  background-color: transparent;
  border: none;
  cursor: pointer;
  height: 24px;
  margin: ${({
    marginLeft,
    marginRight,
  }) => (
    `0px ${marginRight}px 0px ${marginLeft}px`
  )};
  position: relative;
  width: 24px;
`;

Container.defaultProps = {
  marginLeft: 0,
  marginRight: 0,
};

const InnerContainer = styled.div`
  display: flex;
  height: 100%;
  justify-content: center;
  left: 0;
  position: absolute;
  right: 0;
  &.fade-enter {
    opacity: 0;
  }
  &.fade-enter-active {
    opacity: 1;
    transition: 500ms;
  }
  &.fade-exit {
    opacity: 1;
  }
  &.fade-exit-active {
    opacity: 0;
    transition: 500ms;
  }
`;

const Pictogram = ({
  hoverPictogram: HoverPictogramComponent,
  marginLeft = 0,
  marginRight = 0,
  pictogram: PictogramComponent,
}: PictogramI) => {
  const [hover, setHover] = React.useState<boolean>(false);
  return (
    <Container
      marginLeft={marginLeft}
      marginRight={marginRight}
      onMouseEnter={() => setHover(true)}
      onMouseLeave={() => setHover(false)}
    >
      <CSSTransition
        classNames='fade'
        in={hover}
        timeout={300}
        unmountOnExit
      >
        <InnerContainer>
          <HoverPictogramComponent />
        </InnerContainer>
      </CSSTransition>
      <CSSTransition
        classNames='fade'
        in={!hover}
        timeout={300}
        unmountOnExit
      >
        <InnerContainer>
          <PictogramComponent />
        </InnerContainer>
      </CSSTransition>
    </Container>
  );
};

export default Pictogram;
