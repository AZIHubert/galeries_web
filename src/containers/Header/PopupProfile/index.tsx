import * as React from 'react';

import Modal from '#components/Modal';

const PopupProfile = () => {
  const [openModal, setOpenModal] = React.useState<boolean>(false);
  const handleClose = () => {
    setOpenModal(false);
  };
  return (
    <>
      <div
        data-testid='popupProfil'
      >
        <button>
          <img
            alt='profile picture'
          />
          <p>
              userName
          </p>
          <p>
              see your profile
          </p>
        </button>
        <button>
          <img
            alt='edit your information pictogram'
          />
          <p>
              Edit your informations
          </p>
        </button>
        <button
          data-testid='buttonTicketModal'
          onClick={() => setOpenModal((nextState) => !nextState)}
        >
          <img
            alt='send a ticket pictogram'
          />
          <p>
              Share your opinion? Find a bug?
          </p>
          <p>
              send a ticket
          </p>
        </button>
      </div>
      <Modal
        open={openModal}
        handleClose={handleClose}
      >
        <div
          data-testid='ticketModal'
        />
      </Modal>
    </>
  );
};

export default PopupProfile;
