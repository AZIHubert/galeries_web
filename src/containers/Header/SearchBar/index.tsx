import * as React from 'react';

import search from '#ressources/svg/search.svg';

import {
  Container,
  Input,
  Logotype,
} from './styles';

const SearchBar = () => {
  const [searchBarIsFocused, setSearchBarIsFocused] = React.useState<boolean>(false);
  const searchInputRef = React.useRef<HTMLInputElement | null>(null);
  return (
    <Container
      focused={searchBarIsFocused}
      onClick={() => {
        if (searchInputRef.current) {
          searchInputRef.current.focus();
        }
      }}
    >
      <Logotype
        alt='search pictogram'
        focused={searchBarIsFocused}
        src={search}
      />
      <Input
        onBlur={() => setSearchBarIsFocused(false)}
        onFocus={() => setSearchBarIsFocused(true)}
        placeholder={searchBarIsFocused ? '' : 'search on Galeries'}
        ref={searchInputRef}
      />
    </Container>
  );
};

export default SearchBar;
