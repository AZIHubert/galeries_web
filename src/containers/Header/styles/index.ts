export { default as Container } from './Container';
export { default as HeaderPart } from './HeaderPart';
export { default as InnerContainer } from './InnerContainer';
export { default as Logo } from './Logo';
