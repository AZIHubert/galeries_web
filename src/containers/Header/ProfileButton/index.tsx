import * as React from 'react';
import styled from 'styled-components';
import { CSSTransition } from 'react-transition-group';

import defaultProfilePicture from '#ressources/svg/defaultProfilePicture.svg';

import PopupProfile from './PopupProfile';

const Container = styled.div`
  position: relative;
`;

const Button = styled.button`
  align-items: center;
  background-color: transparent;
  border: none;
  border-left: ${({ theme }) => (
    `1px solid ${theme.colors.primary}`
  )};
  border-right: ${({ theme }) => (
    `1px solid ${theme.colors.primary}`
  )};
  color: ${({ theme }) => theme.colors.primary};
  cursor: pointer;
  display: flex;
  font-size: 1.1rem;
  padding: 0 30px;
  &:focus {
    outline: none;
  }
`;

const ProfileButtonImage = styled.img`
  border: ${({ theme }) => (
    `2px solid ${theme.colors.primary}`
  )};
  border-radius: 50%;
  height: 36px;
  margin-right: 20px;
  width: 36px;
`;

const ProfileButton = () => {
  const [openPopupProfil, setOpenPopupProfile] = React.useState<boolean>(false);
  const containerRef = React.useRef<HTMLDivElement | null>(null);
  const handleClickOutside = (event: any) => {
    if (containerRef.current && !containerRef.current.contains(event.target)) {
      setOpenPopupProfile(false);
    }
  };
  React.useEffect(() => {
    document.addEventListener('click', handleClickOutside, true);
    return () => {
      document.removeEventListener('click', handleClickOutside, true);
    };
  });
  return (
    <Container
      ref={containerRef}
    >
      <Button
        data-testid='profilButton'
        onClick={() => setOpenPopupProfile((lastState) => !lastState)}
      >
        <ProfileButtonImage
          alt='profile picture'
          src={defaultProfilePicture}
        />
          userName
      </Button>
      <CSSTransition
        classNames='fade'
        in={openPopupProfil}
        timeout={500}
        unmountOnExit
      >
        <PopupProfile />
      </CSSTransition>
    </Container>
  );
};

export default ProfileButton;
