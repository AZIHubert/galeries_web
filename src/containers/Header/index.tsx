import * as React from 'react';

import logo from '#ressources/svg/logoG.svg';

import PopupProfile from './PopupProfile';

const Header = () => {
  const [openPopupProfil, setOpenPopupProfile] = React.useState<boolean>(false);
  return (
    <header>
      <div>
        <img src={logo} alt="header logo" />
        <img
          alt='search pictogram'
        />
        <input
          placeholder='search on Galeries'
        />
      </div>
      <button>
        <img
          alt='home pictogram'
        />
      </button>
      <button>
        <img
          alt='create new galerie pictogram'
        />
      </button>
      <button>
        <img
          alt='notification pictogram'
        />
      </button>
      <button
        data-testid='profilButton'
        onClick={() => setOpenPopupProfile((lastState) => !lastState)}
      >
        <img
          alt='profile picture'
        />
        userName
      </button>
      {openPopupProfil && <PopupProfile />}
      <button>
        logout
      </button>
    </header>
  );
};

export default Header;
