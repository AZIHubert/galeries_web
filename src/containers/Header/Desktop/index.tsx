import * as React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import {
  useHistory,
  useLocation,
} from 'react-router-dom';
import { CSSTransition } from 'react-transition-group';

import Button from '#components/Button';
import Modal from '#components/Modal';

import logo from '#ressources/svg/logoG.svg';
import {
  CreateGalerie,
  CreateGalerieHover,
  Home,
  HomeHover,
  Notification,
  NotificationHover,
} from '#ressources/svgComponents';

import { fetchLogout } from '#store/actions';
import { userSelector } from '#store/selectors';

import CreateGalerieModal from './CreateGalerieModal';
import Pictogram from './Pictogram';
import ProfileButton from './ProfileButton';
import {
  Container,
  Fader,
  HeaderPart,
  InnerContainer,
  Logo,
  StyledLink,
} from './styles';

const Header = () => {
  const dispatch = useDispatch();
  const history = useHistory();
  const user = useSelector(userSelector);
  const location = useLocation();
  const handleClickHome = () => history.push('/');
  const [openCreateGalerie, setOpenCreateGalerie] = React.useState<boolean>(false);

  const handleOpenCreateGalerie = () => setOpenCreateGalerie(true);
  const handleCloseCreateGalerie = () => setOpenCreateGalerie(false);

  const show = !!user
    && !location.pathname.includes('profilePicture')
    && !location.pathname.includes('updateEmail')
    && !location.pathname.includes('validateEmail');

  return (
    <CSSTransition
      classNames='fade'
      in={show}
      timeout={300}
      unmountOnExit
    >
      <Fader>
        <Container>
          <InnerContainer>
            <HeaderPart>
              <StyledLink
                to='/'
              >
                <Logo
                  alt="header logo"
                  src={logo}
                />
              </StyledLink>
            </HeaderPart>
            <HeaderPart>
              <Pictogram
                hoverPictogram={HomeHover}
                marginRight={25}
                marginRightL={35}
                onClick={handleClickHome}
                pictogram={Home}
              />
              <Pictogram
                hoverPictogram={CreateGalerieHover}
                marginRight={25}
                marginRightL={35}
                onClick={handleOpenCreateGalerie}
                pictogram={CreateGalerie}
              />
              <Pictogram
                hoverPictogram={NotificationHover}
                pictogram={Notification}
              />
              <ProfileButton />
              <Button.Header
                onClick={() => dispatch(fetchLogout())}
                small
                styles={{
                  marginLeft: 40,
                }}
                title='logout'
              />
            </HeaderPart>
          </InnerContainer>
          <Modal.Portal
            handleClose={handleCloseCreateGalerie}
            open={openCreateGalerie}
          >
            <CreateGalerieModal
              handleCloseCreateGalerie={handleCloseCreateGalerie}
            />
          </Modal.Portal>
        </Container>
      </Fader>
    </CSSTransition>
  );
};

export default Header;
