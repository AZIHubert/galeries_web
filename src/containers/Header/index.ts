import Desktop from './Desktop';
import Home from './Home';

const Header = {
  Desktop,
  Home,
};

export default Header;
