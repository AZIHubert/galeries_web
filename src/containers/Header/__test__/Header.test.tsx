import * as React from 'react';
import renderer from 'react-test-renderer';
import {
  cleanup,
  fireEvent,
  render,
  screen,
} from '@testing-library/react';

import Header from '../index';

describe('Header', () => {
  afterEach(cleanup);
  it('renders without crashing', () => {
    const tree = renderer.create(<Header />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should open popup profile', async () => {
    const { getByTestId } = render(<Header />);
    const profilButton = getByTestId('profilButton');
    fireEvent.click(profilButton);
    const popupProfil = await screen.findByTestId('popupProfil');
    expect(popupProfil).not.toBeNull();
  });
  it('should close popup profile', async () => {
    const { getByTestId } = render(<Header />);
    const profilButton = getByTestId('profilButton');
    fireEvent.click(profilButton);
    fireEvent.click(profilButton);
    const popupProfil = screen.queryByTestId('popupProfil');
    expect(popupProfil).toBeNull();
  });
});
