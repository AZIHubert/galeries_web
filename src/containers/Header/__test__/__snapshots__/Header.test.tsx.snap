// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders without crashing 1`] = `
<header>
  <div>
    <img
      alt="header logo"
      src={Object {}}
    />
    <img
      alt="search pictogram"
    />
    <input
      placeholder="search on Galeries"
    />
  </div>
  <button>
    <img
      alt="home pictogram"
    />
  </button>
  <button>
    <img
      alt="create new galerie pictogram"
    />
  </button>
  <button>
    <img
      alt="notification pictogram"
    />
  </button>
  <button
    data-testid="profilButton"
    onClick={[Function]}
  >
    <img
      alt="profile picture"
    />
    userName
  </button>
  <button>
    logout
  </button>
</header>
`;
